test/k8sT/Services.go

```diff
@@ -152,6 +167,89 @@ var _ = Describe("K8sServicesTest", func() {
 			}
 		}, 300)
+
+		It("Checks connectivity is correct during deployment scale", func() {
+			// Add a goroutine to check the connectivity meanwhile
+			// service/deployment is scaling or downgrading.
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			type BackgroundTestAsserts struct {
+				res  *helpers.CmdRes
+				time time.Time
+			}
+			backgroundChecks := []*BackgroundTestAsserts{}
+			started := make(chan struct{})
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				close(started)
+				for {
+					select {
+					case <-ctx.Done():
+						wg.Done()
+						return
+					default:
+						res := kubectl.ExecPodCmd(
+							helpers.DefaultNamespace, appPods[helpers.App2],
+							helpers.CurlFail("http://%s/public", clusterIP))
+						assert := &BackgroundTestAsserts{
+							res:  res,
+							time: time.Now(),
+						}
+						backgroundChecks = append(backgroundChecks, assert)
+						time.Sleep(backgroundSleepInterval)
+					}
+				}
+			}()
+			<-started
+
+			By("Autoscaling %q to 4 replicas", helpers.App1)
+			res := kubectl.Exec(fmt.Sprintf("%s scale deployment %s --replicas=4",
+				helpers.KubectlCmd, helpers.App1))
+			res.ExpectSuccess("Cannot scale up %s to 4 replicas", helpers.App1)
+
+			err := kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).To(BeNil(), "Pods are not ready after scaling up")
+
+			pods, err := kubectl.GetPodNames(helpers.DefaultNamespace, fmt.Sprintf("id=%s", helpers.App1))
+			Expect(err).To(BeNil(), "Cannot get pods names for app1 deployment")
+			Expect(len(pods)).To(Equal(4), "Mismatch in number of scaled pods")
+
+			By("Autoscaling %q to 2 replicas", helpers.App1)
+
+			res = kubectl.Exec(fmt.Sprintf("%s scale deployment %s --replicas=2",
+				helpers.KubectlCmd, helpers.App1))
+			res.ExpectSuccess("Cannot scale up %s to 2 replicas", helpers.App1)
+
+			ExpectAllPodsTerminated(kubectl)
+
+			pods, err = kubectl.GetPodNames(helpers.DefaultNamespace, fmt.Sprintf("id=%s", helpers.App1))
+			Expect(err).To(BeNil(), "Cannot get pods names for app1 deployment")
+			Expect(len(pods)).To(Equal(2), "Mismatch in number of scaled pods")
+
+			By("Checking connectivity during service scaling")
+			cancel()
+			wg.Wait()
+
+			backgroundConnectionTotal := len(backgroundChecks)
+			backgroundConnectionFailed := 0
+			backgroundConnectionFailedMessage := ""
+			Expect(backgroundChecks).ShouldNot(BeEmpty(), "No background connections were made")
+			for _, check := range backgroundChecks {
+				if !check.res.WasSuccessful() {
+					backgroundConnectionFailed++
+					backgroundConnectionFailedMessage += fmt.Sprintf(
+						"Connection to the service at %s failed \n%s",
+						check.time, check.res.OutputPrettyPrint())
+				}
+			}
+			GinkgoPrint("Made %d connections in total and %d failed",
+				backgroundConnectionTotal, backgroundConnectionFailed)
+			Expect(backgroundConnectionFailed).ShouldNot(
```

Comment:
Add a comment indicating why we are allowing at most `maxNumberOfBackgroundConnectionFails` connection failures. This is because we do not want to break the CI immediately with these changes. Is there a way you can output if any connection failure occurred as a warning?

Reply:
fixed.